datasync/chaindatafetcher/api.go

```diff
@@ -51,6 +56,12 @@ func (api *PublicChainDataFetcherAPI) ReadCheckpoint() (int64, error) {
 }
 
 func (api *PublicChainDataFetcherAPI) WriteCheckpoint(checkpoint int64) error {
+	isRunning := atomic.LoadUint32(&api.f.fetchingStarted)
+	if isRunning == running {
+		return errors.New("please call stopFetching first before write checkpoint manually")
+	}
+
+	api.f.checkpoint = checkpoint
```

Comment:
Is there any possibility to concurrent write of `checkpoint`?

Reply:
@ehnuje `checkpoint` is updated only when `fetching` is started which is checked the above statement. Apart from that, it is not possible to access `checkpoint` concurrently.